datafusion/sql/src/planner.rs

```diff
@@ -2678,9 +2683,13 @@ pub fn convert_data_type(sql_type: &SQLDataType) -> Result<DataType> {
 
 // Parse number in sql string, convert to Expr::Literal
 fn parse_sql_number(n: &str) -> Result<Expr> {
-    match n.parse::<i64>() {
-        Ok(n) => Ok(lit(n)),
-        Err(_) => Ok(lit(n.parse::<f64>().unwrap())),
+    match (n.parse::<i64>(), n.parse::<f64>()) {
+        (Ok(n), _) => Ok(lit(n)),
+        (Err(_), Ok(n)) => Ok(lit(n)),
+        (Err(_), Err(_)) => Err(DataFusionError::from(ParserError(format!(
+            "Cannot parse {} as i64 or f64",
+            n
+        )))),
     }
```

Comment:
It may be simpler to do this in two steps?

```suggestion
    match n.parse::<i64>() {
        Ok(n) => Ok(lit(n)),
        _ => match n.parse::<f64>() {
            Ok(n) => Ok(lit(n)),
            _ => Err(DataFusionError::from(ParserError(format!("Cannot parse {} as i64 or f64", n)))),
        }
    }
```

Reply:
Here is an alternate proposal of (yet) another way to do it: https://github.com/apache/arrow-datafusion/pull/3432